python/dolfinx/mesh.py

```diff
@@ -111,19 +104,85 @@ def create_mesh(comm, cells, x, domain, ghost_mode=cpp.mesh.GhostMode.shared_fac
     return mesh
 
 
-def MeshTags(mesh, dim, indices, values):
+_meshtags_types = {
+    numpy.int8: cpp.mesh.MeshTags_int8,
+    numpy.int32: cpp.mesh.MeshTags_int32,
+    numpy.int64: cpp.mesh.MeshTags_int64,
+    numpy.double: cpp.mesh.MeshTags_double
+}
+
 
-    if isinstance(values, int):
-        values = numpy.full(indices.shape, values, dtype=numpy.int32)
-    elif isinstance(values, float):
-        values = numpy.full(indices.shape, values, dtype=numpy.double)
+class MeshTags:
+    def __init__(self, *args):
+        """This initializer is not intended for the user interface. Use create_mesh_tags"""
+        try:
+            mesh, dim, indices, values = args
+            if isinstance(values, int):
+                values = numpy.full(indices.shape, values, dtype=numpy.intc)
+            elif isinstance(values, float):
+                values = numpy.full(indices.shape, values, dtype=numpy.double)
+            dtype = values.dtype.type
+            try:
+                self._cpp_object = _meshtags_types[dtype](mesh, dim, indices, values)
+            except KeyError:
+                raise KeyError("Unsupported value type ({}) for MeshTags.".format(dtype))
+        except ValueError:
+            assert len(args) == 1
+            self._cpp_object = args[0]
+
+    @property
+    def dim(self):
+        return self._cpp_object.dim
+
+    @property
+    def mesh(self):
+        return self._cpp_object.mesh
+
+    @property
+    def name(self):
+        return self._cpp_object.name
+
+    @name.setter
+    def name(self, val):
+        self._cpp_object.name = val
+
+    @property
+    def indices(self):
+        return self._cpp_object.indices
+
+    @property
+    def values(self):
+        return self._cpp_object.values
+
+    def ufl_id(self):
+        return self._cpp_object.id
+
+
+def create_meshtags(mesh: cpp.mesh.Mesh, dim: int, indices: typing.Union[numpy.array, cpp.graph.AdjacencyList_int32],
```

Comment:
I find it confusing to merge 2 very different code paths into seemingly one method.
1. A code path where I pass directly indices of mesh entities to pass is a very fast simple assignment in C++.
2. If I pass AdjacencyList, i.e. local entity-vertex connectivities this has to do much more work to find actual entity indices.

Moreover C++'s create_meshtags() is reserved for the option 2. We discussed that simple object instantiation which is almost no-op (option 1.) should be done in constructor.